File: openvt/web/js/mapper.js
import { OneEuroFilter } from './filter.js';

const HEAD = new Set(['headX', 'headY', 'headAngleZ']);
const clamp = (v, lo, hi) => Math.max(lo, Math.min(hi, v));

export class Mapper {
  constructor(cfg = {}){
    this.mirror     = cfg.mirror !== false;
    this.yawSign    = cfg.yaw_sign   ?? 1;
    this.pitchSign  = cfg.pitch_sign ?? 1;
    this.rollSign   = cfg.roll_sign  ?? 1;
    this.headGain   = cfg.head_gain  ?? 1.3;
    this.deadzone   = cfg.deadzone_deg ?? 0.6;
    [this.blinkLo, this.blinkHi] = cfg.blink_range || [0.35, 0.7];
    this.linkEyes   = cfg.link_eyes   ?? 0.3;
    this.headCutoff = cfg.head_cutoff ?? 1.5;
    this.faceCutoff = cfg.face_cutoff ?? 4.0;
    this.beta       = cfg.smooth_beta ?? 0.1;
    this.filters = new Map();
    this.neutral = [0, 0, 0];
    this._calib = false;
  }
  calibrate(){ this._calib = true; }
  _f(id, x, t){
    let f = this.filters.get(id);
    if (!f){ f = new OneEuroFilter(HEAD.has(id) ? this.headCutoff : this.faceCutoff, this.beta);
             this.filters.set(id, f); }
    return f.apply(x, t);
  }
  _open(raw){
    const x = clamp((raw - this.blinkLo) / Math.max(this.blinkHi - this.blinkLo, 1e-3), 0, 1);
    return x * x * (3 - 2 * x);
  }
  _dz(v){ return Math.abs(v) < this.deadzone ? 0 : v - Math.sign(v) * this.deadzone; }

  apply(frame, params, t, skip){
    if (!frame || !frame.ok) return false;
    if (this._calib){
      this.neutral = [frame.yaw, frame.pitch, frame.roll];
      for (const f of this.filters.values()) f.reset();
      this._calib = false;
    }
    const ms = this.mirror ? -1 : 1;
    const yaw = this._dz(frame.yaw - this.neutral[0]);
    const pitch = this._dz(frame.pitch - this.neutral[1]);
    const roll = this._dz(frame.roll - this.neutral[2]);

    const out = {
      headX: yaw * ms * this.yawSign * this.headGain,
      headY: -pitch * this.pitchSign * this.headGain,
      headAngleZ: -roll * ms * this.rollSign,
      mouthOpen: clamp(frame.jawOpen * 1.6, 0, 1),
      mouthForm: clamp(frame.smile * 1.8 - frame.pucker * 1.5, -1, 1),
      browY: clamp((frame.browUp - frame.browDown) * 1.6, -1, 1),
      eyeBallX: clamp(-ms * (frame.lookR - frame.lookL) * 1.5, -1, 1),
      eyeBallY: clamp((frame.lookD - frame.lookU) * 1.5, -1, 1),
    };
    let oL = this._open(1 - frame.blinkL), oR = this._open(1 - frame.blinkR);
    if (Math.abs(oL - oR) < this.linkEyes){ oL = oR = (oL + oR) / 2; }
    out.eyeOpenL = this.mirror ? oL : oR;
    out.eyeOpenR = this.mirror ? oR : oL;

    for (const [id, v] of Object.entries(out)){
      if (skip.has(id) || !params.has(id)) continue;
      params.set(id, this._f(id, v, t));
    }
    return true;
  }
}

export class IdleAnimator {
  constructor(){ this.next = 2; this.t0 = null; }
  apply(params, t, tracked, skip){
    if (params.has('breath') && !skip.has('breath'))
      params.set('breath', 0.5 - 0.5 * Math.cos(t * 2 * Math.PI / 3.6));
    if (tracked) return;
    const idle = {
      headX: 5 * Math.sin(t * 0.6), headY: 2.5 * Math.sin(t * 0.9 + 1),
      headAngleZ: 2 * Math.sin(t * 0.45), eyeBallX: 0.25 * Math.sin(t * 0.3)
    };
    for (const [id, v] of Object.entries(idle)) if (!skip.has(id)) params.set(id, v);
    if (this.t0 === null && t >= this.next) this.t0 = t;
    let o = 1;
    if (this.t0 !== null){
      const u = (t - this.t0) / 0.22;
      if (u >= 1){ this.t0 = null; this.next = t + 2 + Math.random() * 3; }
      else o = Math.abs(u * 2 - 1);
    }
    for (const id of ['eyeOpenL', 'eyeOpenR']) if (!skip.has(id)) params.set(id, o);
  }
}

File: openvt/track/mapper.py
from __future__ import annotations
import math
import random
from core.filter import OneEuroFilter

HEAD_PARAMS = {"headX", "headY", "headAngleZ"}


def _clamp(v, lo, hi):
    return max(lo, min(hi, v))


class Mapper:
    """TrackingFrame -> params. All signs/gains live here, not in the model."""

    def __init__(self, cfg: dict | None = None):
        cfg = cfg or {}
        self.mirror = bool(cfg.get("mirror", True))
        self.yaw_sign = float(cfg.get("yaw_sign", 1.0))
        self.pitch_sign = float(cfg.get("pitch_sign", 1.0))
        self.roll_sign = float(cfg.get("roll_sign", 1.0))
        self.head_gain = float(cfg.get("head_gain", 1.3))
        self.deadzone = float(cfg.get("deadzone_deg", 0.6))
        self.blink_lo, self.blink_hi = cfg.get("blink_range", [0.35, 0.7])
        self.link_eyes = float(cfg.get("link_eyes", 0.3))
        self.head_cutoff = float(cfg.get("head_cutoff", 1.5))
        self.face_cutoff = float(cfg.get("face_cutoff", 4.0))
        self.beta = float(cfg.get("smooth_beta", 0.1))
        self.filters: dict[str, OneEuroFilter] = {}
        self.neutral = [0.0, 0.0, 0.0]
        self._calib = False

    def calibrate(self):
        self._calib = True

    def _f(self, pid, x, t):
        f = self.filters.get(pid)
        if f is None:
            cutoff = self.head_cutoff if pid in HEAD_PARAMS else self.face_cutoff
            f = self.filters[pid] = OneEuroFilter(cutoff, self.beta)
        return f(x, t)

    def _open(self, raw_open):
        x = _clamp((raw_open - self.blink_lo) / max(self.blink_hi - self.blink_lo, 1e-3), 0.0, 1.0)
        return x * x * (3 - 2 * x)

    def _dz(self, v):
        return 0.0 if abs(v) < self.deadzone else v - math.copysign(self.deadzone, v)

    def apply(self, frame, params, t, skip=()) -> bool:
        if frame is None or not frame.ok:
            return False
        if self._calib:
            self.neutral = [frame.yaw, frame.pitch, frame.roll]
            for f in self.filters.values():
                f.reset()
            self._calib = False

        ms = -1.0 if self.mirror else 1.0
        yaw = self._dz(frame.yaw - self.neutral[0])
        pitch = self._dz(frame.pitch - self.neutral[1])
        roll = self._dz(frame.roll - self.neutral[2])

        out = {
            "headX": yaw * ms * self.yaw_sign * self.head_gain,
            "headY": -pitch * self.pitch_sign * self.head_gain,
            "headAngleZ": -roll * ms * self.roll_sign,
            "mouthOpen": _clamp(frame.jaw_open * 1.6, 0.0, 1.0),
            "mouthForm": _clamp(frame.smile * 1.8 - frame.pucker * 1.5, -1.0, 1.0),
            "browY": _clamp((frame.brow_up - frame.brow_down) * 1.6, -1.0, 1.0),
            "eyeBallX": _clamp(-ms * (frame.look_r - frame.look_l) * 1.5, -1.0, 1.0),
            "eyeBallY": _clamp((frame.look_d - frame.look_u) * 1.5, -1.0, 1.0),
        }
        oL, oR = self._open(1 - frame.blink_l), self._open(1 - frame.blink_r)
        if abs(oL - oR) < self.link_eyes:
            oL = oR = (oL + oR) / 2
        if self.mirror:
            out["eyeOpenL"], out["eyeOpenR"] = oL, oR
        else:
            out["eyeOpenL"], out["eyeOpenR"] = oR, oL

        for pid, v in out.items():
            if pid in skip or not params.has(pid):
                continue
            params.set(pid, self._f(pid, v, t))
        return True


class IdleAnimator:
    """Breathing always; gentle sway + auto blink when nothing is tracking."""

    def __init__(self):
        self.next_blink = 2.0
        self.blink_t0 = None

    def apply(self, params, t, dt, tracked: bool, skip=()):
        if params.has("breath") and "breath" not in skip:
            params.set("breath", 0.5 - 0.5 * math.cos(t * 2 * math.pi / 3.6))
        if tracked:
            return
        idle = {"headX": 5 * math.sin(t * 0.6), "headY": 2.5 * math.sin(t * 0.9 + 1),
                "headAngleZ": 2 * math.sin(t * 0.45), "eyeBallX": 0.25 * math.sin(t * 0.3)}
        for pid, v in idle.items():
            if pid not in skip:
                params.set(pid, v)
        if self.blink_t0 is None and t >= self.next_blink:
            self.blink_t0 = t
        o = 1.0
        if self.blink_t0 is not None:
            u = (t - self.blink_t0) / 0.22
            if u >= 1.0:
                self.blink_t0 = None
                self.next_blink = t + random.uniform(2.0, 5.0)
            else:
                o = abs(u * 2 - 1)
        for pid in ("eyeOpenL", "eyeOpenR"):
            if pid not in skip:
                params.set(pid, o)